agent/csi/volumes.go

```diff
@@ -0,0 +1,259 @@
+package csi
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/docker/swarmkit/agent/exec"
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/log"
+)
+
+// volumes is a map that keeps all the currently available volumes to the agent
+// mapped by volume ID.
+type volumes struct {
+	mu        sync.RWMutex                     // To sync between Add(), Remove() and Reset() for "m" map
+	tryMu     sync.RWMutex                     // To sync between tryAddVolume() and tryRemoveVolume()
+	mapMu     sync.RWMutex                     // To sync access to "pluginMap" map
+	m         map[string]*api.VolumeAssignment // Map between VolumeID and VolumeAssignment
+	pluginMap map[string]*NodePlugin           // Map between Driver Name and NodePlugin
+}
+
+const maxRetries int = 25
+
+const initialBackoff = 100 * time.Millisecond
+
+// NewManager returns a place to store volumes.
+func NewManager() exec.VolumesManager {
+	return &volumes{
+		m:         make(map[string]*api.VolumeAssignment),
+		pluginMap: make(map[string]*NodePlugin),
+	}
+}
+
+// Get returns a volume published path for the provided volume ID.  If the volume doesn't exist, returns empty string.
+func (r *volumes) Get(volumeID string) (string, error) {
+	r.mapMu.RLock()
+	defer r.mapMu.RUnlock()
+	if plugin, ok := r.pluginMap[volumeID]; ok {
+		return plugin.GetPublisedPath(volumeID)
+	}
+	return "", nil
+}
+
+// Add adds one or more volumes to the volume map.
+func (r *volumes) Add(volumes ...api.VolumeAssignment) {
+	r.mu.Lock()
+	var volumeObjects []*api.VolumeAssignment
+	defer r.mu.Unlock()
+	ctx := context.Background()
+	for _, volume := range volumes {
+		v := volume.Copy()
+		log.G(ctx).WithField("method", "(*volumes).Add").Debugf("Add Volume:%v", volume.VolumeID)
+
+		r.m[volume.VolumeID] = v
+		driverName := v.Driver.Name
+		r.mapMu.Lock()
+		if _, ok := r.pluginMap[driverName]; !ok {
+			r.pluginMap[driverName] = NewNodePlugin(driverName)
+		}
+		r.mapMu.Unlock()
+		volumeObjects = append(volumeObjects, v)
+	}
+	go r.iterateVolumes(volumeObjects, true)
```

Comment:
This goroutine is unnecessary, because the function it executes just spawns its own goroutines.